src/lib.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/// YkMir -- Metadata serialiser and deserialiser for Yorick.
+///
+/// This crate allows ykrustc to serialise the compiler's MIR structures for later deserialisation
+/// by the Yorick runtime. The encoded data is specialised to a particular Rust input program (this
```

Comment:
What does "specialised to a particular Rust input" mean?

Reply:
OK, I only have a vague understand of this at the current time, but as I understand it:

The metadata encoded into rlibs and dynlibs by Rustc is "generic" so that things like monomorphisations, type resolution and inlining can happen for arbitrary inputs (Rust programs) in later compilation sessions.

When we read *our* metadata at runtime, we want the data for the current binary. We don't care about the general case. We have two options:

 * Spin up the compiler to resolve rusts metadata to our binary (not even sure how possible this is).
 * Resolve everything at compile time and serialise that.

I'm proposing the latter.

Reply:
We could just drop the paragraph about specialisation until we understand it better.

I admit, there's a liberal amount of hand waving happening here.

Reply:
I think it might be better to drop it if we don't understand it :) That way we leave ourselves a bit of flexibility in the future if we haven't quite got things right / fully understood them.